Sources/Sentry/SentryAppStartTracker.m

```diff
@@ -0,0 +1,170 @@
+    static dispatch_once_t once;
+    [self.dispatchQueue
+        dispatchOnce:&once
+               block:^{
+                   NSString *appStartType = [self getStartType];
+
+                   if ([appStartType isEqualToString:SentryAppStartTypeUnkown]) {
+                       [SentryLog logWithMessage:@"Unknown start type. Not measuring app start."
+                                        andLevel:kSentryLevelWarning];
+                   } else if (self.wasInBackground) {
+                       // If the app was already running in the background it's not a cold or warm
+                       // start.
+                       [SentryLog logWithMessage:@"App was in background. Not measuring app start."
+                                        andLevel:kSentryLevelInfo];
+                   } else {
+                       // According to a talk at WWDC about optimizing app launch (
+                       // https://devstreaming-cdn.apple.com/videos/wwdc/2019/423lzf3qsjedrzivc7/423/423_optimizing_app_launch.pdf?dl=1
+                       // slide 17) no process exists for cold and warm launches. Therefore it is
+                       // fine to use the process start timestamp. Instead on Android the process
+                       // can be forked before the app is launched and this would give wrong values.
+                       // Using the proess start time returned valid values when testing with real
+                       // devices.
+                       // It could be that we have to switch back to setting a appStart-timestamp in
+                       // the load method of this class to get a close approximation of when the
+                       // process was started.
+                       NSTimeInterval appStartTime = [[self.currentDate date]
+                           timeIntervalSinceDate:self.sysctl.processStartTimestamp];
+
+                       [SentryLog logWithMessage:[NSString stringWithFormat:@"AppStart: sys:%f",
+                                                           appStartTime]
+                                        andLevel:kSentryLevelInfo];
+
+                       SentryAppStartMeasurement *appStartMeasurement =
+                           [[SentryAppStartMeasurement alloc] initWithType:appStartType
+                                                                  duration:appStartTime];
+                       SentrySDK.appStartMeasurement = appStartMeasurement;
+                   }
+               }];
+
+    [self stop];
+}
+
+- (NSString *)getStartType
+{
+    // App launched the first time
+    if (self.previousAppState == nil) {
+        return SentryAppStartTypeCold;
+    }
+
+    SentryAppState *currentAppState = [self.appStateManager buildCurrentAppState];
+
+    // If the release name is different we assume it's an app upgrade
+    if (![currentAppState.releaseName isEqualToString:self.previousAppState.releaseName]) {
+        return SentryAppStartTypeCold;
+    }
```

Comment:
L: why do we need to account for that? do we rely on something stored in the disk to distinguish app state?

Reply:
We store the app state to disk yes. We already do this for [OOM](https://github.com/getsentry/sentry-cocoa/blob/8b8e68c58cdc3f7ba63b8304c57d7fa73841b809/Sources/Sentry/SentryOutOfMemoryLogic.m#L53).